### PR TITLE
Akka Streams Kafka 0.11

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
Please review @adamw @kciesielski

cc @myfear @drewhk @2m @johanandren @ktoso

**Comment by kciesielski** [comment]
Looks good :+1:

**Comment by myfear** [comment]
LGTM! Would love to also see a Java example if possible?

File: _posts/2016-09-10-akka-stream-kafka.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+author: Adam Warski
+category: integrations
+series_title: Integration
+series_tag: integration
+tags: [streams,integration]
+---
+{% include JB/setup %}
+
+
+[Apache Kafka](http://kafka.apache.org) is the leading distributed messaging system, and [Reactive Streams](http://www.reactive-streams.org) is an emerging standard for asynchronous stream processing. It seems natural to combine these two; that’s why [SoftwareMill](https://softwaremill.com) started the [reactive-kafka](https://github.com/akka/reactive-kafka) project back in December 2014 and maintained it since.
```
Comment (ktoso) [inline _posts/2016-09-10-akka-stream-kafka.md:14]: Would it not be good to explain that apps often use Kafka WITH Akka in the apps?
I often get confused people who ask "Kafka vs. Akka"-like questions.

File: _posts/2016-09-10-akka-stream-kafka.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+Here’s a couple of common Kafka usage scenarios, comparing the old reactive-kafka version (M4), the current version (0.11), and equivalent functionality implemented using plain Kafka `Producer`s/`Consumer`s (but not taking into account, of course, connecting with any other reactive components).
+
+![Plain consumer]({{ site.url }}/assets/kafka-bench1.png)
+
+The first scenario represents a consumer which reads messages from Kafka and pushes it through a non-blocking processing stage, without commit. The reactive setup (akka-plain-consumer) gets very close (80%) to full speed achieved with a while loop pulling elements from a consumer (plain-consumer).
+
+![Consumer with batched commit]({{ site.url }}/assets/kafka-bench2.png)
+
+Another scenario represents a very common use case - consuming messages and committing in batches in order to achieve at-least once delivery. Performance improvements in 0.11 allowed to gain a massive throughput increase from ~17k to ~450k msgs/s.
```
Comment (ktoso) [inline _posts/2016-09-10-akka-stream-kafka.md:55]: Should we mention on what box this was measured?

File: _posts/2016-09-10-akka-stream-kafka.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+![At most once consumer]({{ site.url }}/assets/kafka-bench3.png)
+
+Sometimes one needs to commit each single message before processing, which gives at most once delivery guarantee. 0.11 optimizations fixed the super-slow value of 20 msgs/s and reached over 2800 msgs/s. This is a great step forward, very close to the simple while loop scenario (“at-most-once”).
```
Comment (ktoso) [inline _posts/2016-09-10-akka-stream-kafka.md]: 20 msgs/s _in the previous implementation_, and ...

**Comment by patriknw** [comment]
@myfear We can link to the documentation, which contains similar examples in java http://doc.akka.io/docs/akka-stream-kafka/current/home.html

File: _posts/2016-09-10-akka-stream-kafka.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    (batch, elem) => batch.updated(elem) })
+  .mapAsync(3)(_.commitScaladsl())
+  .runWith(Sink.ignore)
+
+result.onFailure {
+  case e: Throwable => handleError(e)
+}
+```
+
+Secondly, there are significant performance improvements. We’ve done some benchmarks, and while there is still some overhead from the reactive-kafka wrapper (though keep in mind, that you also get something in return: e.g. all of the benefits coming from backpressure!), the overall numbers look very good and are getting better.
```
Comment (ktoso) [inline _posts/2016-09-10-akka-stream-kafka.md]: Can we add that it also is a great API, much easier to work with than the native kafka one?

File: _posts/2016-09-10-akka-stream-kafka.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+Sometimes one needs to commit each single message before processing, which gives at most once delivery guarantee. 0.11 optimizations fixed the super-slow value of 20 msgs/s and reached over 2800 msgs/s. This is a great step forward, very close to the simple while loop scenario (“at-most-once”).
+
+![Producer]({{ site.url }}/assets/kafka-bench4.png)
+
+Our last benchmark tests a producer continuously sending messages to a Kafka topic. Current version of akka-stream-kafka writes over 85,000 messages per second. This result has been achieved with producer parallelism set to 100, which indicates how many parallel writes can be waiting for confirmation until this stage backpressures. The “plain-producer” test chosen for comparison writes to Kafka in a loop being bound only by its internal buffer size.
+
+Give it a try, we’d love to hear your feedback!
+
+
```
Comment (ktoso) [inline _posts/2016-09-10-akka-stream-kafka.md:67]: @adamw @kciesielski just FYI, I guess it's too late to fix that... The diagrams look not as good, they make one focus that we're slower, and forget about the benefits.
Competition is saying random things and numbers claiming they're "faster than native", which ofc is not true, but from a selling the awesomeness of the lib perspective we're not doing too well here.

I guess it does not matter as much, anyone who tried to use the nativ APIs knows it's a pain and will like the Akka Streams abstraction

**Comment by ktoso** [comment]
I didn't see the graphs before in the doc where we edited this - I don't think they're that good "marketing wise" (well, it is a release announcement thus it is a marketing thing).
But anyway, let's ship it - it's much better than it used to be after all!

LGTM

File: _posts/2016-09-10-akka-stream-kafka.md (hunk 4 above)
Comment (patriknw) [reply _posts/2016-09-10-akka-stream-kafka.md]: we can say "benefits coming from a nice API with asynchronous backpressure"

Comment (ktoso) [reply _posts/2016-09-10-akka-stream-kafka.md]: that's great yes

**Comment by patriknw** [comment]
@kciesielski I added you as author also

Publishing it now

**Comment by kciesielski** [comment]
Thanks @patriknw

**Comment by patriknw** [comment]
it's live http://blog.akka.io/integrations/2016/09/10/akka-stream-kafka

**Comment by kciesielski** [comment]
Just noticed a typo in `Softwaremill`, should be `SoftwareMill` :)

**Pull request merged**